### ISSUE TITLE
Loss function for "mostly implicit" data

### ISSUE THREAD

**Issue by christinedekock** [lyst/lightfm#310]
Hi Maciej,

I am using LightFM to train a recommender model on data from a content serving platform. The vast majority of the training data (about 9m events, 65k users, 35k items) is implicit - ie. pageviews. However, I also have a small amount of dislike data, where users could indicate that they wanted to 'See fewer posts like this'. This only makes up about 5k events, but it's obviously a pretty strong signal and I would like to include it somehow. I tried training a model with the logistic loss function and (+1/-1) scores as suggested, but it performed very poorly - presumably because we have almost no explicitly negative events. It seems to be doing better with the WARP loss function. Does it make sense at all to use WARP instead? And is there anything else you would recommend I change or try?

I am using 256 components (established through a grid search), with all other hyperparameters at their default values. Early stopping is used. 

The scores for **log loss** are as follows on the test set:
- **precision**:0.001725
- **recall:** 0.00427
- **auc**: 0.79170
- **reciprocal rank**: 0.02387

For **WARP**:
- **precision**: 0.00728
- **recall**: 0.01340
- **auc**: 0.80441
- **reciprocal rank**: 0.03838

Any feedback would be greatly appreciated. Thank you for all your work on this package!

**Comment by reditya** [comment]
Hi @christinedekock,

Recently, I also explored similar situation as yours, where in addition to positive implicit feedback data, I also have implicit negative feedback. The way I approach the problem is instead of sampling negative items at random, I modified a bit the sampling strategy to sample from implicit negative feedback. This is somewhat intuitive IMO, because if we already have the negative feedback for each users, why would we choose at random? This is even more relevant if we have sufficient number of negative feedbacks. 

Comparing to random sampling, this small tweak lead to better MAP with fewer epoch. I think it helps the model to converge faster (in terms of epoch). I'm still performing extensive evaluation though, but the initial result seems promising. I uses the WARP loss, haven't tried using logistic loss though.

I'd like to say thanks as well to @maciejkula  for this great package!

**Comment by maciejkula** [comment]
@reditya your approach makes sense! The main reason why something like this isn't in LightFM is that it's somewhat tricky to make sure we cover this case as well as the more common case where we just want uniform negative sampling (and where we don't confuse novice users!).

**Comment by maciejkula** [comment]
@christinedekock an alternative approach is to train two WARP models, one for positive and one for negative interactions, and ensemble them together.

**Comment by theoaf** [comment]
Hi thanks for the great package @maciejkula 

do you have any suggestions for an explicit feedback recommender? 

any packages? any examples?
